### PR TITLE
Fix invalid module bitcode when indexing a swizzled bool vector

### PR DESCRIPTION
**Pull request by amaiorano**
When indexing a swizzled bool vector, some HLSL-specific code in EmitCXXMemberOrOperatorMemberCallExpr kicks in to handle the HLSLVecType. In this case, we’re dealing with an ExtVectorElt because of the swizzle, so this function creates a GEP, Load, and Store on the vector. However, boolean scalars are returned as type i11 while the store is storing to a bool, which is an i32, so we need to insert a cast before the store.

### CODE REVIEW

**Comment by amaiorano** [comment]
See this example of the failure: https://godbolt.org/z/WsajscbT9

File: tools/clang/test/CodeGenDXIL/operators/swizzle/indexSwizzledBoolVec.hlsl
Comment (llvm-beanz) [inline tools/clang/test/CodeGenDXIL/operators/swizzle/indexSwizzledBoolVec.hlsl:7]: Checking that it doesn't crash isn't a good enough test. We tend to expect that inputs both don't crash _and_ produce valid output.

You should be able to add the `-fcgl` flag to get the pre-optimized IR, and do some checks against that to verify that the IR generation is correct.

File: tools/clang/lib/CodeGen/CGExprCXX.cpp
Comment (llvm-beanz) [inline tools/clang/lib/CodeGen/CGExprCXX.cpp:238]: nit:
```suggestion
        Type *ElTy = This->getType()->getPointerElementType()
                                    ->getVectorElementType();
        if (llvm::IntegerType *IT = dyn_cast<llvm::IntegerType>(ThisTy))
          if (IT->getBitWidth() != 1)
            return Val;

        // bool pointer is not i1 *.
        llvm::Type *BoolTy = llvm::IntegerType::get(
            getLLVMContext(), getContext().getTypeSize(CE->getType()));
        Val = Builder.CreateBitCast(
            Val, llvm::PointerType::get(
                     BoolTy, Val->getType()->getPointerAddressSpace()));
        return Val;
```

Comment (amaiorano) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:238]: My original patch actually looked like this, but I wanted to make it clear that I had simply refactored the existing code mostly as-is. Will apply this change.

File: tools/clang/test/CodeGenDXIL/operators/swizzle/indexSwizzledBoolVec.hlsl
Comment (amaiorano) [reply tools/clang/test/CodeGenDXIL/operators/swizzle/indexSwizzledBoolVec.hlsl:7]: That's fair. I had based this on other examples I saw of hlsl tests that seem to validate that dxc doesn't crash. But a better test is a good thing, so I'll do this.

File: tools/clang/lib/CodeGen/CGExprCXX.cpp
Comment (amaiorano) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp]: Uploaded new patch based on our conversation.

File: tools/clang/test/CodeGenDXIL/operators/swizzle/indexSwizzledBoolVec.hlsl
Comment (amaiorano) [reply tools/clang/test/CodeGenDXIL/operators/swizzle/indexSwizzledBoolVec.hlsl]: Done.

File: tools/clang/lib/CodeGen/CGExpr.cpp
Comment (llvm-beanz) [inline tools/clang/lib/CodeGen/CGExpr.cpp:1245]: Just as a comment for any additional reviewers. This matches what modern clang does. Basically we can skip inserting value range metadata for bools that are represented as i1 because the valid range of a bool is the full range of an i1.

**Review by llvm-beanz — COMMENTED** [review]
I think this looks good, but since I contributed a bit to it I want to get some additional eyes on it.

**Comment by farzonl** [comment]
LGTM

**Review by farzonl — APPROVED** [review]

File: tools/clang/lib/CodeGen/CGExpr.cpp
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGExpr.cpp:1140]: Shouldn't this be static?  Plus the name is generic, but seems to only apply to bool?  Is that correct?

File: tools/clang/lib/CodeGen/CGExpr.cpp
Comment (tex3d) [reply tools/clang/lib/CodeGen/CGExpr.cpp:1245]: Is bool the only type we would want to emit range information for?  `ShouldEmitRangeMD` won't return true unless hasBooleanRepresentation(Ty) returns true AND has bit size > 1.  It isn't obvious to me that this is correct.

File: tools/clang/lib/CodeGen/CGExprCXX.cpp
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: Will this impact conversion of bool that could come from host memory?  Trunc is only valid if you know that the source is already a bit pattern constrained by the compiler.  Otherwise, we must use comparison with zero (not-equal) to translate from i32 in storage to i1.

**Review by hekota — APPROVED** [review]
LGTM!

File: tools/clang/test/CodeGenDXIL/operators/swizzle/indexSwizzledBoolVec.hlsl
Comment (tex3d) [inline tools/clang/test/CodeGenDXIL/operators/swizzle/indexSwizzledBoolVec.hlsl:11]: `trunc` is ok if original bool was not host-visible and guaranteed to be a known bit representation, but what if this `v_bool4` array is in host-visible memory?  In that case we have to convert with comparisons against zero.  Are we sure that path works?  We should have a test for that case, I think.

File: tools/clang/lib/CodeGen/CGExpr.cpp
Comment (llvm-beanz) [reply tools/clang/lib/CodeGen/CGExpr.cpp:1140]: Good catch!

File: tools/clang/lib/CodeGen/CGExpr.cpp
Comment (llvm-beanz) [inline tools/clang/lib/CodeGen/CGExpr.cpp:1142]:
```suggestion
  if(hasBooleanRepresentation(Ty))
    return cast<llvm::IntegerType>(Value->getType())->getBitWidth() != 1;
  return true;
```

File: tools/clang/lib/CodeGen/CGExpr.cpp
Comment (llvm-beanz) [reply tools/clang/lib/CodeGen/CGExpr.cpp:1140]:
```suggestion
static bool ShouldEmitRangeMD(llvm::Value *Value, QualType Ty) {
```

File: tools/clang/lib/CodeGen/CGExpr.cpp
Comment (llvm-beanz) [reply tools/clang/lib/CodeGen/CGExpr.cpp:1245]: That’s a bug in the code I gave @amaiorano. I’ve posted suggestions to correct.

File: tools/clang/lib/CodeGen/CGExprCXX.cpp
Comment (llvm-beanz) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: It is undefined behavior in C/C++ and HLSL for bool to have any value other than 0 or 1. Given that, trunc is valid in all cases converting between lengths of bool.

Comment (amaiorano) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: I've applied Chris' fixes based on your comments @tex3d. I'm waiting for your approval before merging.

Comment (llvm-beanz) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: To fully cite the spec. This is the wording in the C++ spec:

> Values of type bool are either true or false.
[HTML Citation](https://timsong-cpp.github.io/cppwp/n4140/basic.fundamental#6)

Since HLSL defines `true` to `1` and `false` to `0` (the same as pretty much all C/C++ compilers), any different value is UB.

Comment (tex3d) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: While I can't find documentation to back this up now, HLSL compilers (FXC and DXC) have always required a translation for bools coming from host visible memory (cbuffer, etc...) by interpreting any non-zero value as true, and zero value as false.  It's something we did in FXC, and had to do in DXC to match FXC.  Trunc will convert any bit pattern with the lowest bit unset to `0`.  This would be inconsistent with the way bools coming from host memory have been handled thus far for HLSL.

Comment (tex3d) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: For one thing, DXBC instructions are defined in a way that an incoming bool value from a constant buffer may be used in an `if` instruction, and the `if` instruction interprets the 32-bit value with any bit set as `true`, and no bits set as `false`.
See the instruction spec [here](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#22.7.1%20Branch%20based%20on%20boolean%20condition:%20if_condition).

Comment (tex3d) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: Though, as luck would have it, under `Restrictions:` `2)` it gets `_z` and `_nz` modifiers exactly backwards.  The earlier description of the operation gets it right though.

Comment (llvm-beanz) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: I don’t see how the behavior of the DXBC instruction is relevant here. This is absolutely not a behavior you can rely on LLVM to match. Clang (and by extension DXC) have been assuming that the only valid values for bool are 0 and 1 (see the bit where this PR changes the range check).

If a bool value comes in from somewhere and it is anything other than 0 or 1, it’s going to break in a lot of places. The advantage to using a `trunc` here is that it is _way_ easier for LLVM to optimize and eliminate unnecessary `trunc`, `zext` and `sext` instructions because the don’t always depend on value analysis.

If you feel strongly enough about this to hold up the PR we should have a meeting to discuss.

Comment (llvm-beanz) [reply tools/clang/lib/CodeGen/CGExprCXX.cpp:248]: We had a lengthy discussion internally and I think we all went back and forth about different thoughts here. In practice since the existing code causes a crash, and this causes correct behavior where the only possible error cases are very firmly in the UB, I think we should take the change as-is.

If @tex3d feels strongly about this we can discuss this more and revisit.

There are a lot of broken behaviors around HLSL boolean vectors that we're going to need to revisit and reconsider in our Clang implementation and we can discuss longer term solutions in that context.

**Review by llvm-beanz — APPROVED** [review]

**Pull request merged**
